### PR TITLE
Aseprite v1.3 file spec changes

### PR DESCRIPTION
**Pull request by B-Reif**
# v1.3 File Spec Changes

This PR adds new features to support [the upcoming changes to the Aseprite file spec.](https://gist.github.com/dacap/35f3b54fbcd021d099e0166a4f295bab). This involves both additional features and refactors of existing code. I'll document changes in the order provided by the spec, and then append changes specific to the crate.

### Tile data type
From the spec:
> `TILE`: **Tilemaps**: Each tile can be a 8-bit (`BYTE`), 16-bit (`WORD`), or 32-bit (`DWORD`) value and there are masks related to the meaning of each bit.

[The new `Tile` struct in `tile.rs`](https://github.com/alpine-alpaca/asefile/pull/2/files#diff-a3eb4f70f7764a3618ed357b37f5241814aa35a18049df3feeac39a4d240bafcR1) represents this type. Currently, Aseprite itself only supports 32-bit `TILE` values, so this PR only targets 32-bit values as well.
### Tilemap layer type
[The new `Tilemap` LayerType enum case](https://github.com/alpine-alpaca/asefile/pull/2/files#diff-5eab58a7f7ef848629a8be1084b12464ced845ebd69087b9cbc1a2263465212aR20) stores a Tileset index (`DWORD`)  value.
### Tilemap cel
From the spec:
> For cel type = 3 (Compressed Tilemap)
      WORD      Width in number of tiles
      WORD      Height in number of tiles
      WORD      Bits per tile (at the moment it's always 32-bit per tile)
      DWORD     Bitmask for tile ID (e.g. 0x1fffffff for 32-bit tiles)
      DWORD     Bitmask for X flip
      DWORD     Bitmask for Y flip
      DWORD     Bitmask for 90CW rotation
      BYTE[10]  Reserved
      TILE[]    Row by row, from top to bottom tile by tile
                compressed with ZLIB method (see NOTE.3)

[The new `Tilemap` struct stores this data](https://github.com/alpine-alpaca/asefile/pull/2/files#diff-ee79f45f1e3779841e4dc4150977b612a835d179c015070f94d43602a16ed0bdR1). The `Cel` struct has been refactored to support [a new `Tilemap` enum case](https://github.com/alpine-alpaca/asefile/pull/2/files#diff-8b1884e1c3cf6aded7f25be068a12022ea2c60a1631dfb422e08400004b7e4daR277).
### External files chunk
From the spec:
> A list of external files linked with this file. It might be used to
reference external palettes or tilesets.
    DWORD       Number of entries
    BYTE[8]     Reserved (set to zero)
    For each entry
      DWORD     Entry ID (this ID is referenced by tilesets or palettes)
      BYTE[8]   Reserved (set to zero)
      STRING    External file name

[The new `external_file` module](url) supports parsing these chunks and the `AsepriteFile` struct now exposes them. Aseprite itself currently has no way of creating .aseprite files with these chunks, so for now we have no test cases against this.

### User data chunk changes
From the spec:
>After the tags chunk, you can write one user data chunk for each tag. E.g. if there are 10 tags, you can then write 10 user data chunks one for each tag.
...
In version 1.3 a sprite has associated user data, to consider this case there is an User Data Chunk at the first frame after the Palette Chunk.

Asefile currently just parses UserData chunks and discards them. To support these chunks properly, we'll need to also contextualize them (e.g. Sprite UserData, Tag UserData, etc). This PR does not include any changes around this.

### Tileset chunk
[Link to the spec](https://gist.github.com/dacap/35f3b54fbcd021d099e0166a4f295bab#file-ase-diff-L145)

[The new `tileset` module](https://github.com/alpine-alpaca/asefile/pull/2/files#diff-996471aa880231940e2975bcd5b1f9501f9ee0b3c476e8505d73e945d912231eR1) supports creating these chunks and exposing them as part of the exported AsepriteFile struct.

### AseReader
[The new AseReader struct](https://github.com/alpine-alpaca/asefile/pull/2/files#diff-267c9f4da66412a9f439ac08d224356fe24265b5e1cebb6c44c2d55b96414513R1) wraps over some common reading behaviors. This change touches most lines of 'input' code to use AseReader methods.
### Pixel
[The new pixel module](https://github.com/alpine-alpaca/asefile/pull/2/files#diff-bc06baf899772fdd6d8fb86e7901e22d37aa9382e329d3c139ce2aa9efef2139R1) supports mapping the various pixel formats of Aseprite to structs. This supports image logic for both Tilesets and Cels, which each contain pixel data.
### Tests
[Two new test cases](https://github.com/alpine-alpaca/asefile/pull/2/files#diff-a0262fea108af44a98cc49e5eb72641963dd816513f2d0a22eb738fcfed55ebeR329) test against RGBA and Indexed Tilemap files respectively. We can add more tests (and functionality) once Aseprite itself catches up to its file spec.

### CODE REVIEW

File: src/layer.rs [hunk 1]
```diff
@@ -123,6 +126,7 @@ pub struct LayerData {
     pub(crate) blend_mode: BlendMode,
     pub(crate) opacity: u8,
     pub(crate) layer_type: LayerType,
+    pub(crate) tilemap_index: Option<u32>,
```
Comment (B-Reif) [inline src/layer.rs:129]: It might be better to put this data under the `LayerType::Tilemap` enum. In that case the Layer struct would need a little more refactoring.

**Comment by alpine-alpaca** [comment]
Looking good. Would be important to have a few test files, though. I haven't yet figured out how to install Aseprite 1.3 without disabling my existing 1.2 version.

**Comment by B-Reif** [comment]
Still a bit more to do to support External File and Tileset chunks. I'll be sure to include some test files.

File: src/reader.rs [hunk 2]
```diff
@@ -0,0 +1,71 @@
+use crate::{AsepriteParseError, Result};
+use byteorder::{LittleEndian, ReadBytesExt};
+use flate2::read::ZlibDecoder;
+use std::io::{Cursor, Read, Seek, SeekFrom};
+
+pub(crate) struct AseReader<T: Read + Seek> {
```
Comment (B-Reif) [inline src/reader.rs:6]: This new struct `AseReader` wraps over some common reading behaviors:

- Bakes in `<LittleEndian>` so we don't have to specify it everywhere
- Uses methods named after the terms in the Aseprite file spec (WORD, DWORD, etc)
- Exposes some methods that were previously done in helper functions like reading strings, skipping bytes, and filling a buffer.

This change requires the underlying type to be both `Read + Seek` to support the skip function. If that's a problem, we can change `skip` to a Read only implementation.

Comment (alpine-alpaca) [reply src/reader.rs:6]: Yes, makes sense. It's a nice cleanup.

In the current usage it doesn't even need to be generic. It could just wrap a `Cursor`, because every section is read into a `Vec<u8>` first. So we can't have a fully streaming implementation, but I'm not planning to add this until someone has a very compelling use case.

One thing I'm thinking about is to keep image data compressed in memory, as that might potentially save a lot of RAM. But that's a more complicated task and would have to be a separate PR.

**Comment by B-Reif** [comment]
Tilesets can contain the actual image data (in the spec as `PIXEL[]`). This means some actual pixel-parsing logic needs to be refactored out of the Cel module and into some shared Pixel module.

**Comment by B-Reif** [comment]
I don't exactly understand how to parse a Tilemap cel. It has [a list of 'tiles'](https://gist.github.com/dacap/35f3b54fbcd021d099e0166a4f295bab#file-ase-diff-L76), and [a 'tile' is a value](https://gist.github.com/dacap/35f3b54fbcd021d099e0166a4f295bab#file-ase-diff-L11). I'm not really sure how to interpret the Tile value to get pixel data. do you have any insight on this? @alpine-alpaca

**Comment by alpine-alpaca** [comment]
> 
> 
> I don't exactly understand how to parse a Tilemap cel. It has [a list of 'tiles'](https://gist.github.com/dacap/35f3b54fbcd021d099e0166a4f295bab#file-ase-diff-L76), and [a 'tile' is a value](https://gist.github.com/dacap/35f3b54fbcd021d099e0166a4f295bab#file-ase-diff-L11). I'm not really sure how to interpret the Tile value to get pixel data. do you have any insight on this? @alpine-alpaca

My guess is that the tile size divides the cel into regions of tile width x tile height. And then the image data to be used is indicated by the tile id (ie. `TILE`). And the actual data is stored in the Tileset chunk, which stores its data as an image of width tile width and height (tiles x tile height). So the cel data is just an array of ids. and it should be be of size (`ceil(image width / tile width) * ceil(image height / tile height)`.

**Comment by B-Reif** [comment]
Okay, I've marked this as Ready and updated the OP with a quick tour of the PR. Tests are passing, including the new Tilemap tests.

File: src/tileset.rs [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+    /// Number to show in the UI for the tile with index=0. Default is 1.
+    /// Only used for Aseprite UI purposes. Not used for data representation.
+    pub fn base_index(&self) -> &i16 {
+        &self.base_index
+    }
+    pub fn name(&self) -> &String {
+        &self.name
+    }
+    /// When Some, includes a link to an external file.
+    pub fn external_file(&self) -> &Option<ExternalTilesetReference> {
```
Comment (alpine-alpaca) [inline src/tileset.rs:114]: I think this should return a `Option<&...>`, i.e., call `.as_ref()` in the body.

File: src/tileset.rs [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+        &self.external_file
+    }
+    /// TODO: Tileset image export
+    pub fn image(&self) -> RgbaImage {
+        todo!()
```
Comment (alpine-alpaca) [inline src/tileset.rs:119]: Don't leave a `todo!` in here. Just remove/uncomment the whole function if you want to do it in a separate PR.

File: src/tileset.rs [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+    }
+}
+impl Index<TilesetId> for TilesetsById {
+    type Output = Tileset;
+    fn index(&self, id: TilesetId) -> &Self::Output {
+        let map = self.map();
+        if map.contains_key(&id) {
+            return &self.map()[&id];
+        }
+        panic!("no external file found for id")
```
Comment (alpine-alpaca) [inline src/tileset.rs:187]: Hm, again not a fan of this panic. I guess in this case, it's possible for the user to avoid it? I think it would be better to just provide a `get` method and not an `Index` impl. Then the user has to decide whether to call `unwrap` or not.

File: src/cel.rs [hunk 6]
```diff
@@ -131,37 +133,44 @@ impl CelsData {
...
-                CelData::RawIndexed { .. } => {
-                    return Err(AsepriteParseError::InvalidInput("Internal error: unresolved Indexed data".into()));
+        if let Some(ref cel) = layers[layer_index] {
+            match &cel.content {
+                CelContent::Raw(image_content) => {
+                    match image_content.pixels {
+                        Pixels::Rgba(_) => {
+                            // TODO: Verify data length
+                        }
+                        Pixels::Grayscale(_) => todo!(),
```
Comment (alpine-alpaca) [inline src/cel.rs:145]: No crashes/panics please. Return an "unsupported feature" error instead.

File: src/external_file.rs [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+    }
+}
+impl Index<ExternalFileId> for ExternalFilesById {
+    type Output = ExternalFile;
+    fn index(&self, id: ExternalFileId) -> &Self::Output {
+        let map = self.map();
+        if map.contains_key(&id) {
+            return &self.map()[&id];
+        }
+        panic!("no external file found for id")
```
Comment (alpine-alpaca) [inline src/external_file.rs:69]: Again, not a fan of panics. It would be safer to just provide a `get` method.

File: src/file.rs [hunk 8]
```diff
@@ -58,6 +64,14 @@ impl PixelFormat {
             PixelFormat::Indexed { .. } => 1,
         }
     }
+    pub(crate) fn expect_indexed(&self) -> &u8 {
+        match self {
+            PixelFormat::Indexed {
+                transparent_color_index,
+            } => transparent_color_index,
+            _ => panic!("Expected an Indexed PixelFormat"),
```
Comment (alpine-alpaca) [inline src/file.rs:72]: Another panic. That should probably be turned into a `Result`.

File: src/file.rs [hunk 9]
```diff
@@ -205,71 +238,92 @@ impl AsepriteFile {
...
-            self.copy_cel(&mut image, cel);
+            self.write_cel(&mut image, cel);
         }
 
         image
     }
 
-    fn copy_cel(&self, image: &mut RgbaImage, cel: &RawCel) {
+    fn write_cel(&self, image: &mut RgbaImage, cel: &RawCel) {
         assert!(self.pixel_format != PixelFormat::Grayscale);
```
Comment (alpine-alpaca) [inline src/file.rs:248]: Ok, I see this function contains a bunch of asserts and panics. The reasoning was that all those invariants should be caught by `validate` so they are actually impossible for users. Still we have to double check that that's actually the case.

File: src/file.rs [hunk 10]
```diff
@@ -205,71 +238,92 @@ impl AsepriteFile {
...
-                    &blend_fn,
-                );
+            CelContent::Tilemap(tilemap_data) => {
+                let layer_type = layer.layer_type();
+                let tileset_id = layer_type.expect_tilemap();
+                let tileset = &self.tilesets()[*tileset_id];
+                let tileset_pixels = tileset
+                    .pixels
+                    .as_ref()
+                    .expect("Expected Tileset data to contain pixels");
```
Comment (alpine-alpaca) [inline src/file.rs:274]: For all these `expect` calls: They either need to be verified by the validator. Or we should print a `warn!` and just do nothing instead. But we shouldn't crash/panic.

File: src/file.rs [hunk 11]
```diff
@@ -355,28 +409,83 @@ fn blend_mode_to_blend_fn(mode: BlendMode) -> BlendFn {
...
+    // let pixels = tileset
+    //     .pixels()
+    //     .as_ref()
+    //     .expect("Tileset missing tile pixel data")
+    //     .expect_rgba();
+    let blend_fn = blend_mode_to_blend_fn(*blend_mode);
+
+    for tile_y in 0..tilemap_height {
+        for tile_x in 0..tilemap_width {
+            // TODO: support tile transform flags
```
Comment (alpine-alpaca) [inline src/file.rs:452]: Shouldn't be too difficult since they are just transformations on the pixel coordinates.
Also the `image` function provides them as operators so if we find a solution for the blend function stuff we could use those.

File: src/layer.rs [hunk 12]
```diff
@@ -11,6 +15,17 @@ pub enum LayerType {
...
     Group,
+    /// A tilemap layer. Contains the tileset index.
+    /// In Aseprite these are represented by a grid icon.
+    Tilemap(TilesetId),
+}
+impl LayerType {
+    pub(crate) fn expect_tilemap(&self) -> &TilesetId {
+        match self {
+            Self::Tilemap(id) => id,
+            _ => panic!("Expected a Tilemap layer"),
```
Comment (alpine-alpaca) [inline src/layer.rs:26]: Another panic. Should either be turn into a `Result` or made impossible by validating first.

File: src/pixel.rs [hunk 13]
```diff
@@ -0,0 +1,155 @@
...
+}
+impl Pixels {
+    fn from_bytes(bytes: Vec<u8>, pixel_format: PixelFormat) -> Result<Self> {
+        match pixel_format {
+            PixelFormat::Indexed { .. } => {
+                let pixels = bytes.iter().map(|byte| Indexed(*byte)).collect();
+                Ok(Self::Indexed(pixels))
+            }
+            PixelFormat::Grayscale => {
+                assert!(bytes.len() % 2 == 0);
```
Comment (alpine-alpaca) [inline src/pixel.rs:90]: Turn into `Err` please.

File: src/pixel.rs [hunk 14]
```diff
@@ -0,0 +1,155 @@
...
+                let pixels: Result<Vec<_>> = bytes.chunks_exact(2).map(Grayscale::new).collect();
+                pixels.map(Self::Grayscale)
+            }
+            PixelFormat::Rgba => {
+                assert!(bytes.len() % 4 == 0);
```
Comment (alpine-alpaca) [inline src/pixel.rs:95]: Also, turn to `Err`.

**Review by alpine-alpaca — CHANGES_REQUESTED** [review]
Ok. Thanks for your efforts. Overall I like it, and the reader and pixel abstractions make the code more readable and clean.

The biggest issue right now are the `assert`s and other ways to panic. Since this is a library that should handle arbitrary data, we don't ever want to crash the user's program. We have three strategies to handle issues of unmet expectations:
* Return an error for functions that return a `Result`.
* Use validation (after parsing) so that later functions can rely on those invariants.
* For expectations that should be handled by the validator but aren't (i.e., bugs) it might still be best not to panic, but to abort the operation and log a warning or an error. This means the operation will produce an incorrect result, but at least it won't crash the whole program.

File: src/tile.rs [hunk 15]
```diff
@@ -0,0 +1,62 @@
...
+impl Index<usize> for Tiles {
+    type Output = Tile;
+
+    fn index(&self, index: usize) -> &Self::Output {
+        &self.0[index]
+    }
+}
+
+fn as_bool(bit: u32) -> bool {
+    assert!(bit == 0 || bit == 1);
```
[the remark this answers is not in the document]
Comment (B-Reif) [reply src/tile.rs:60]: Okay, we can check for 'nonzero' instead of '1'.

The Aseprite UI doesn't support this feature yet, so if we want to create test cases we'll need to write our own test files.

File: src/tile.rs [hunk 16]
```diff
@@ -0,0 +1,62 @@
...
+
+#[derive(Debug)]
+pub(crate) struct TileId(pub(crate) u32);
+
+#[derive(Debug)]
+pub(crate) struct Tile {
+    pub id: TileId,
+    pub flip_x: bool,
+    pub flip_y: bool,
+    pub rotate_90cw: bool,
```
[the remark this answers is not in the document]
Comment (B-Reif) [reply src/tile.rs:15]: Yeah, that's the idea. You'd get different results if the 90cw rotation is applied _before_ or _after_ the flips, so when Aseprite implements this feature we'll need to test to make sure we do whatever it does.

File: src/tileset.rs [hunk 17]
```diff
@@ -0,0 +1,189 @@
...
+    }
+}
+bitflags! {
+    struct TilesetFlags: u32 {
+        /// Include link to external file.
+        const LINKS_EXTERNAL_FILE = 0x0001;
+        /// Include tiles inside this file.
+        const FILE_INCLUDES_TILES = 0x0002;
+        /// Tilemaps using this tileset use tile ID=0 as empty tile
+        /// (this is the new format). In rare cases this bit is off,
```
[the remark this answers is not in the document]
Comment (B-Reif) [reply src/tileset.rs:29]: This flags struct is just for internal use. I'll change these from the doc comment to just a double slash, and update the doc comment on the Tileset struct.

File: src/tile.rs [hunk 18]
```diff
@@ -0,0 +1,62 @@
...
+        header: &TileBitmaskHeader,
+    ) -> Result<Self> {
+        // Only 32-bit tiles supported for now
+        let expected_output_size = 4 * expected_tile_count;
+        let bytes = reader.unzip(expected_output_size)?;
+        let tiles: Result<Vec<Tile>> = bytes
+            .chunks_exact(4)
+            .map(|bytes| Tile::new(bytes, header))
+            .collect();
+        tiles.map(Self)
```
[the remark this answers is not in the document]
Comment (B-Reif) [reply src/tile.rs:48]: The annotation is telling `.collect` to convert from `Iter<Result<Tile>>` to `Result<Vec<Tile>>`. I'll replace this with `Ok(Tiles(tiles?))`

File: src/file.rs (hunk 8 above)
Comment (B-Reif) [reply src/file.rs:72]: I was just using this internally in the `write_cel` function. I'll remove this and make it clearer that this should be caught by validation.

**Comment by B-Reif** [comment]
Ok, I've changed all of the panics in functions that already return `Result`. We can either refactor/extend the validation logic upstream, or we can change the `file` module functions to return a `Result<()>`.

File: src/file.rs (hunk 11 above)
Comment (B-Reif) [reply src/file.rs:455]: Since Aseprite doesn't support these yet, I don't know what order the transforms should be applied in. We could ask David I guess :)

**Comment by B-Reif** [comment]
Pushed a commit to return `Result<()>` from `.image` methods. We can revert this for something else if we want to pursue a different strategy.

**Comment by alpine-alpaca** [comment]
Yes, in earlier versions I had `image` return a result, but then changed it because I felt it's difficult to deal with errors at that stage and it's much nicer to check the assumptions/invariants during parsing where errors must be expected anyway.

**Comment by alpine-alpaca** [comment]
> 
> 
> We can either refactor/extend the validation logic upstream, or we can change the `file` module functions to return a `Result<()>`.

I'm not sure what you mean about `upstream`. Do you mean Aseprite? That wouldn't be enough. This library shouldn't make any assumptions without explicitly validating them. The file could be corrupted in arbitrary ways (intentionally or not) and this library should not crash (just loading the file should return an error). Before a version 1.0 I would like to extend the CI to do some fuzz testing to ensure that it is the case.

**Comment by B-Reif** [comment]
Sorry I was talking about refactoring in either `validate()` or in `image()`, that's all.

**Comment by B-Reif** [comment]
A few new changes:
- added validation functions to several structs
- the parse_info is moved into a function that invokes all of these and returns `Result<ValidatedParseInfo>`
- the expect/panic calls in `.image()` now specify which validation functions it's expecting to have been called

**Comment by B-Reif** [comment]
oops didn't mean to press Close

**Comment by alpine-alpaca** [comment]
Ok, getting there. Some things I noticed:

* `TilesetsById` needs to be exported and documented. It's part of the public interface.
* There's a bunch of clippy warnings. There's some in code you didn't touch, you can ignore those. I'll fix those later.

File: src/file.rs [hunk 19]
```diff
@@ -322,43 +243,96 @@ impl AsepriteFile {
...
+                match tileset_pixels {
+                    pixel::Pixels::Rgba(pixels) => write_tilemap_cel_to_image(
+                        image,
+                        data,
+                        tilemap_data,
+                        tileset,
+                        pixels,
+                        &blend_mode,
+                    ),
+                    pixel::Pixels::Grayscale(_) => todo!(),
```
[the remark this answers is not in the document]
Comment (B-Reif) [reply src/file.rs:282]: >     * There's a bunch of clippy warnings. There's some in code you didn't touch, you can ignore those. I'll fix those later.
Oh, for some reason these weren't appearing in my editor. I'll review them

[the remark this answers is not in the document]
Comment (B-Reif) [reply src/file.rs:282]: I think we can do this

**Comment by B-Reif** [comment]
> Similarly, `TilesetId` probably needs to be exported. Not clear to me ATM how users are meant to construct them. 

My thinking was that the underlying `u32` would be abstracted away and users would only use these Ids for lookup. I added a public constructor in case there's a need for it.

**Comment by B-Reif** [comment]
As an alternative to the current validation logic, we could introduce a separate set of "resolved" data types. Asefile already does resolves indexed pixels to rgba before returning the `AsepriteFile`. We could also resolve tilemap data at that point. Then, we would use "raw" types in `ParseInfo` before resolving them to the types used in `AsepriteFile`.

By statically typing these differences, we could remove all of the `.expect` calls from `.image()` and still have a guaranteed, non-`Result` Image return type. The downside to this is that it's destructive. We lose information when we resolve data before giving it to the user.

**Comment by B-Reif** [comment]
We might be able to further reduce the number of `expect` calls with some changes to the type system, but they would require substantial refactoring and probably need a separate PR. The validate functions should work for now. Let me know if there's anything else you'd like changed.

**Comment by alpine-alpaca** [comment]
Yes, I think we can make things more robust by encapsulating the preconditions/invariants into types some more. Aka ["Parse, don't validate"](https://lexi-lambda.github.io/blog/2019/11/05/parse-don-t-validate/) or ["Make illegal states unrepresentable"](https://blog.janestreet.com/effective-ml-revisited/).
